datafusion/common/src/tree_node.rs

```diff
@@ -432,6 +373,45 @@ pub enum TreeNodeRecursion {
     Stop,
 }
 
+impl TreeNodeRecursion {
+    /// Continues visiting nodes with `f` depending on the current [`TreeNodeRecursion`]
+    /// value and the fact that `f` is visiting the current node's children.
+    pub fn visit_children<F: FnOnce() -> Result<TreeNodeRecursion>>(
+        self,
+        f: F,
+    ) -> Result<TreeNodeRecursion> {
+        match self {
+            TreeNodeRecursion::Continue => f(),
+            TreeNodeRecursion::Jump => Ok(TreeNodeRecursion::Continue),
+            TreeNodeRecursion::Stop => Ok(self),
+        }
+    }
+
+    /// Continues visiting nodes with `f` depending on the current [`TreeNodeRecursion`]
+    /// value and the fact that `f` is visiting the current node's sibling.
+    pub fn visit_sibling<F: FnOnce() -> Result<TreeNodeRecursion>>(
+        self,
+        f: F,
+    ) -> Result<TreeNodeRecursion> {
+        match self {
+            TreeNodeRecursion::Continue | TreeNodeRecursion::Jump => f(),
+            TreeNodeRecursion::Stop => Ok(self),
+        }
+    }
+
+    /// Continues visiting nodes with `f` depending on the current [`TreeNodeRecursion`]
+    /// value and the fact that `f` is visiting the current node's parent.
+    pub fn visit_parent<F: FnOnce() -> Result<TreeNodeRecursion>>(
```

Comment:
I was confused at first about the `parent` term here -- I think it might be clearer to call this function  `visit_self` or `visit_node` as it is used to visit the node (potentially) after visiting its children, rather than visiting the parent

Reply:
I'm ok with `visit_self` or `visit_node`. My rationale for the current `visit_children` / `visit_sibling` / `visit_parent` naming was to specify the relative position to the last `tnr`.

This is a bit more detailed example that I will implement in https://github.com/apache/arrow-datafusion/pull/9913 for `LogicalPlan::visit_with_subqueries` and after we fix https://github.com/apache/arrow-datafusion/pull/9913#discussion_r1548538789 (so the subquery plans are treated like additional children):
```
// visit the current node (self)
visitor.f_down(self)?

    // `visit_children()` because last tnr is from the current node (self) and the continuation 
    // closure visits the current node (self)'s subqueries (that we treat as children)
    .visit_children(|| self.apply_subqueries(|c| c.visit_with_subqueries(visitor)))? 
    
    // `visit_sibling()` because last tnr is from visiting the last subquery (that we treated
    // as a child) and continuation closure visits the current node (self)'s children, that are
    // siblings to the subqueries
    .visit_sibling(|| self.apply_children(|s| s.visit_with_subqueries(visitor)))?

    // `visit_parent()` because last tnr is from transforming the last child and continuation
    // cloure visits the current node (self, that is the parent of the last child)
    .visit_parent(|| visitor.f_up(self)) 
```

So I'm ok with renaming `visit_parent` to `visit_self` or `visit_node`, but then `visit_sibling` sounds weird.

Actually, how about `visit_down` / `visit_next` / `visit_up`?

Reply:
I think the current naming is more descriptive than `visit_down` / `visit_next` / `visit_up` -- let's go with what is in this PR for now and we can iterate in the future if needed